src/downloadAzureProject/setupProjectFolder.ts

```diff
@@ -0,0 +1,103 @@
+    const devContainerName: string = getRequiredQueryParameter(parsedQuery, 'devcontainer');
+    const language: string = getRequiredQueryParameter(parsedQuery, 'language');
+    const action: string = getRequiredQueryParameter(parsedQuery, 'action');
+    const downloadAppContent: boolean = action === HandleUriActions.downloadContentAndSetupProject;
+
+    const toBeDeletedFolderPathUri: vscode.Uri = vscode.Uri.joinPath(vsCodeFilePathUri, 'temp');
+
+    try {
+        const functionAppName: string = getNameFromId(resourceId);
+        const downloadFilePath: string = vscode.Uri.joinPath(toBeDeletedFolderPathUri, `${functionAppName}.zip`).fsPath;
+
+        await vscode.window.withProgress({ location: vscode.ProgressLocation.Notification, title: localize('settingUpFunctionAppLocalProjInfoMessage', `Setting up project for function app '${functionAppName}' with language '${language}'.`) }, async () => {
+            if (downloadAppContent) {
+                // NOTE: We don't want to download app content for compiled languages.
+                const slotTreeItem: SlotTreeItemBase | undefined = await ext.tree.findTreeItem(resourceId, { ...context, loadAll: true });
+                const hostKeys: WebSiteManagementModels.HostKeys | undefined = await slotTreeItem?.client.listHostKeys();
+                const defaultHostName: string | undefined = slotTreeItem?.client.defaultHostName;
+
+                if (!!hostKeys && hostKeys.masterKey && defaultHostName) {
+                    const requestOptions: RequestPrepareOptions = {
+                        method: 'GET',
+                        url: `https://${defaultHostName}/admin/functions/download?includeCsproj=true&includeAppSettings=true`,
+                        headers: { 'x-functions-key':  hostKeys.masterKey }
+                    };
+                    await requestUtils.downloadFile(requestOptions, downloadFilePath);
+                } else {
+                    throw new Error(localize('hostInformationNotFound', 'Failed to get host information for the functionApp.'));
+                }
+            }
+
+            const projectFilePathUri: vscode.Uri = vscode.Uri.joinPath(vsCodeFilePathUri, `${functionAppName}`);
+            const projectFilePath: string = projectFilePathUri.fsPath;
+            const devContainerFolderPathUri: vscode.Uri = vscode.Uri.joinPath(projectFilePathUri, '.devcontainer');
+            if (downloadAppContent) {
+                // tslint:disable-next-line: no-unsafe-any
+                await extract(downloadFilePath, { dir: projectFilePath });
+            }
+            await requestUtils.downloadFile(
+                `https://raw.githubusercontent.com/microsoft/vscode-dev-containers/master/containers/${devContainerName}/.devcontainer/devcontainer.json`,
+                vscode.Uri.joinPath(devContainerFolderPathUri, 'devcontainer.json').fsPath
+            );
+            await requestUtils.downloadFile(
```

Comment:
Just to make sure we're on the same page - I'm still not satisfied with the dev container implementation for reasons we've already discussed. My concerns no longer block this PR because of the feature flag, but will need to be addressed before we can remove the feature flag in the future. Putting aside the implementation, I'm also hoping we can get some user feedback on the design of the dev container piece before the flag is removed to help determine if we need it at all.

Reply:
Yes as we discussed in the meeting as well, we are working on getting the API changes to make the devcontainer part better. Also having this deployed behind a feature flag will help us test this end to end once we have our msi/pkg part done and do some user testing as well to get feedback. We will share the feedback with Matt and you.